meta-client/src/test/java/org/apache/helix/metaclient/impl/zk/TestZkMetaClientAsyncOperations.java

```diff
@@ -0,0 +1,231 @@
+package org.apache.helix.metaclient.impl.zk;
+
+/*
+ * Licensed to the Apache Software Foundation (ASF) under one
+ * or more contributor license agreements.  See the NOTICE file
+ * distributed with this work for additional information
+ * regarding copyright ownership.  The ASF licenses this file
+ * to you under the Apache License, Version 2.0 (the
+ * "License"); you may not use this file except in compliance
+ * with the License.  You may obtain a copy of the License at
+ *
+ *     http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing,
+ * software distributed under the License is distributed on an
+ * "AS IS" BASIS, WITHOUT WARRANTIES OR CONDITIONS OF ANY
+ * KIND, either express or implied.  See the License for the
+ * specific language governing permissions and limitations
+ * under the License.
+ */
+
+import java.util.concurrent.CountDownLatch;
+
+import javax.annotation.Nullable;
+
+import org.apache.helix.metaclient.api.AsyncCallback;
+import org.apache.helix.metaclient.api.MetaClientInterface;
+import org.apache.zookeeper.KeeperException;
+import org.testng.Assert;
+import org.testng.annotations.Test;
+
+
+public class TestZkMetaClientAsyncOperations extends ZkMetaClientTestBase {
+
+  static TestAsyncContext[] asyncContext = new TestAsyncContext[1];
+  static final String entryKey = "/TestAsyncEntryKey";
+  static final String nonExistsEntry = "/a/b/c";
+
+  static class TestAsyncContext {
+    int _asyncCallSize;
+    CountDownLatch _countDownLatch;
+    int[] _returnCode;
+    MetaClientInterface.Stat[] _stats;
+    String[] _data;
+
+    TestAsyncContext(int callSize) {
+      _asyncCallSize = callSize;
+      _countDownLatch = new CountDownLatch(callSize);
+      _returnCode = new int[callSize];
+      _stats = new MetaClientInterface.Stat[callSize];
+      _data = new String[callSize];
+    }
+
+    public CountDownLatch getCountDownLatch() {
+      return _countDownLatch;
+    }
+
+    public void countDown() {
+      _countDownLatch.countDown();
+    }
+
+    public int getReturnCode(int idx) {
+      return _returnCode[idx];
+    }
+
+    public MetaClientInterface.Stat getStats(int idx) {
+      return _stats[idx];
+    }
+
+    public String getData(int idx) {
+      return _data[idx];
+    }
+
+    public void setReturnCodeWhenFinished(int idx, int returnCode) {
+      _returnCode[idx] = returnCode;
+    }
+
+    public void setStatWhenFinished(int idx, MetaClientInterface.Stat stat) {
+      _stats[idx] = stat;
+    }
+
+    public void setDataWhenFinished(int idx, String data) {
+      _data[idx] = data;
+    }
+  }
+
+  @Test
+  public void testAsyncCreateSetAndGet() {
+    asyncContext[0] = new TestAsyncContext(2);
+    try (ZkMetaClient<String> zkMetaClient = createZkMetaClient()) {
```

Comment:
Neat : Excellent call on using try-with-resources here. I feel most of our unstable UTs/tests are due to this non-closed zkclients which are being shared across multiple tests.